### PR TITLE
[KBFS-1523] Add diskLimiter implementation with backpressure

### PR DESCRIPTION
**Pull request by akalin-keybase**

### CODE REVIEW

File: libkbfs/tlf_journal.go [hunk 1]
```diff
@@ -62,7 +62,8 @@ func (ca tlfJournalConfigAdapter) usernameGetter() normalizedUsernameGetter {
 	return ca.Config.KBPKI()
 }
 
-const defaultDiskLimitTimeout = 3 * time.Second
+// Should be slightly more than maxDelay in config_local.go.
```
Comment (akalin-keybase) [inline libkbfs/tlf_journal.go]: move this closer to maxDelay

File: libkbfs/backpressure_disk_limiter.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+	backpressureMaxThreshold int64
+	byteLimit                int64
+	maxDelay                 time.Duration
+	delayFn                  func(context.Context, time.Duration) error
+	s                        *kbfssync.Semaphore
+}
+
+var _ diskLimiter = backpressureDiskLimiter{}
+
+func newBackpressureDiskLimiterWithDelayFunction(
```
Comment (akalin-keybase) [inline libkbfs/backpressure_disk_limiter.go]: add comments to functions in this file

File: libkbfs/config_local.go [hunk 3]
```diff
@@ -905,9 +905,12 @@ func (c *ConfigLocal) EnableJournaling(
 	//
 	// TODO: Also keep track of and limit the inode count.
 	var journalDiskLimit int64 = 10 * 1024 * 1024 * 1024
-	// TODO: Use a diskLimiter implementation that applies
-	// backpressure.
-	diskLimitSemaphore := newSemaphoreDiskLimiter(journalDiskLimit)
+	var backpressureMinThreshold = journalDiskLimit / 2
```
Comment (akalin-keybase) [inline libkbfs/config_local.go]: add comments

**Comment by akalin-keybase** [comment]
Still need to do a cleanup/comment pass, but the backpressure logic is ready for review.

File: libkbfs/backpressure_disk_limiter.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+import (
+	"time"
+
+	"github.com/keybase/kbfs/kbfssync"
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+)
+
+// backpressureDiskLimiter is an implementation of diskLimiter that uses
+// a backpressure.
```
Comment (strib) [inline libkbfs/backpressure_disk_limiter.go]: "a backpressure" > "backpressure".

File: libkbfs/backpressure_disk_limiter.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+}
+
+func (s backpressureDiskLimiter) onJournalDisable(journalBytes int64) {
+	if journalBytes > 0 {
+		s.s.Release(journalBytes)
+	}
+}
+
+func (s backpressureDiskLimiter) getDelay() time.Duration {
+	// Slight hack to get backpressure value.
```
Comment (strib) [inline libkbfs/backpressure_disk_limiter.go]: Why not just add getting the count as a first-class semaphore method?

File: libkbfs/backpressure_disk_limiter.go [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+		byteLimit, maxDelay, delayFn, s,
+	}
+}
+
+func defaultDoDelay(ctx context.Context, delay time.Duration) error {
+	if delay == 0 {
+		return nil
+	}
+
+	timer := time.NewTimer(delay)
```
Comment (strib) [inline libkbfs/backpressure_disk_limiter.go]: Logging here will be useful in debugging, I think.

Comment (akalin-keybase) [inline libkbfs/backpressure_disk_limiter.go]: Done. Although I added the logging to before this is called.

File: libkbfs/backpressure_disk_limiter.go (hunk 5 above)
Comment (akalin-keybase) [inline libkbfs/backpressure_disk_limiter.go]: I will, but not in this PR. Added TODO.

File: libkbfs/tlf_journal.go (hunk 1 above)
Comment (akalin-keybase) [inline libkbfs/tlf_journal.go]: Done.

File: libkbfs/backpressure_disk_limiter.go (hunk 2 above)
Comment (akalin-keybase) [inline libkbfs/backpressure_disk_limiter.go]: Done.

File: libkbfs/config_local.go (hunk 3 above)
Comment (akalin-keybase) [inline libkbfs/config_local.go]: Done.

File: libkbfs/backpressure_disk_limiter.go (hunk 4 above)
Comment (akalin-keybase) [inline libkbfs/backpressure_disk_limiter.go]: Done.

**Pull request merged**